File: modules/midi.c
/*
  midi.c -- implement Python midi parser module

  source file of the GNU LilyPond music typesetter

  (c)  2001 Han-Wen Nienhuys <hanwen@cs.uu.nl>
            Jan Nieuwenhuizen <janneke@gnu.org>

*/

/*

python2
import midi
s = open ("s.midi").read ()
midi.parse_track (s)
midi.parse (s)

*/

#include <python2.0/Python.h>

#if 0
int x = 0;
int *track = &x;
#define debug_print(f, args...) fprintf (stderr, "%s:%d: track: %p :" f, __FUNCTION__, __LINE__, *track, ##args)
#else
#define debug_print(f, args...)
#endif

static PyObject *Midi_error;
static PyObject *Midi_warning;

static PyObject *
midi_error (char *s)
{
  PyErr_SetString (Midi_error, s);
  return 0;
}

static PyObject *
midi_warning (char *s)
{
  PyErr_SetString (Midi_warning, s);
  return 0;
}


typedef struct message {
  unsigned char msg;
  char * description;
} message_t;

message_t channelVoiceMessages[] = {
  0x80, "NOTE_OFF",
  0x90, "NOTE_ON",
  0xA0, "POLYPHONIC_KEY_PRESSURE",
  0xB0, "CONTROLLER_CHANGE",
  0xC0, "PROGRAM_CHANGE",
  0xD0, "CHANNEL_KEY_PRESSURE",
  0xE0, "PITCH_BEND",
  0,0
};

message_t channelModeMessages[] = {
  0x78, "ALL_SOUND_OFF",
  0x79, "RESET_ALL_CONTROLLERS",
  0x7A, "LOCAL_CONTROL",
  0x7B, "ALL_NOTES_OFF",
  0x7C, "OMNI_MODE_OFF",
  0x7D, "OMNI_MODE_ON",
  0x7E, "MONO_MODE_ON",
  0x7F, "POLY_MODE_ON",
  0,0
};

message_t metaEvents[] = {
  0x00, "SEQUENCE_NUMBER",
  0x01, "TEXT_EVENT",
  0x02, "COPYRIGHT_NOTICE",
  0x03, "SEQUENCE_TRACK_NAME",
  0x04, "INSTRUMENT_NAME",
  0x05, "LYRIC",
  0x06, "MARKER",
  0x07, "CUE_POINT",
  0x20, "MIDI_CHANNEL_PREFIX",
  0x21, "MIDI_PORT",
  0x2F, "END_OF_TRACK",
  0x51, "SET_TEMPO",
  0x54, "SMTPE_OFFSET",
  0x58, "TIME_SIGNATURE",
  0x59, "KEY_SIGNATURE",
  0x7F, "SEQUENCER_SPECIFIC_META_EVENT",
  0,0
};

unsigned long int
get_number (char * str, char * end_str, int length)
{
  /* # MIDI uses big-endian for everything */
  long sum = 0;
  int i = 0;
  
  for (; i < length; i++)
    sum = (sum << 8) + (unsigned char) str[i];

  debug_print ("%d:\n", sum);
  return sum;
}

unsigned long int
get_variable_length_number (char **str, char * end_str)
{
  long sum = 0;
  int i = 0;
  while (*str < end_str)
    {
      unsigned char x = **str;
      (*str) ++;
      sum = (sum << 7) + (x & 0x7F);
      if (!(x & 0x80))
	break;
    }
  debug_print ("%d:\n", sum);
  return sum;
}

static PyObject *
read_unimplemented_event (char **track, char *end, unsigned long time,
			  unsigned char x, unsigned char z)
{
  debug_print ("%x:%s", z, "unimplemented MIDI event\n");
  *track += 2;
  return Py_BuildValue ("(iii)", z, *((*track) -2), *((*track) -1));
}

PyObject *
read_one_byte (char **track, char *end, unsigned long time,
	       unsigned char x, unsigned char z)
{
  debug_print ("%x:%s", z, "event\n");
  *track += 1;
  return Py_BuildValue ("(ii)", z, *((*track) -1));
}

PyObject *
read_two_bytes (char **track, char *end, unsigned long time,
		unsigned char x, unsigned char z)
{
  debug_print ("%x:%s", z, "event\n");
  *track += 2;
  return Py_BuildValue ("(iii)", z, *((*track) -2), *((*track) -1));
}

PyObject *
read_three_bytes (char **track, char *end, unsigned long time,
		  unsigned char x, unsigned char z)
{
  debug_print ("%x:%s", z, "event\n");
  *track += 3;
  return Py_BuildValue ("(iiii)", z, *((*track) -3), *((*track) -2), *((*track) -1));
}

PyObject *
read_string (char **track, char *end, unsigned long time,
	     unsigned char x, unsigned char z)
{
  unsigned long length = get_variable_length_number (track, end);
  if (length < *track - end)
    debug_print ("%s", "zero length string\n");
  *track += length;
  debug_print ("%x:%s", length, "string\n");
  return Py_BuildValue ("(is)", z, *((*track) -length));
}

typedef PyObject* (*Read_midi_event)
     (char **track, char *end, unsigned long time,
      unsigned char x, unsigned char z);

Read_midi_event read_ff_byte [16] =
{
  read_three_bytes,  //  0
  read_one_byte,     // 10
  read_one_byte,  // 20
  read_one_byte,  // 30
  read_one_byte,  // 40
  read_one_byte,  // 50
  read_one_byte,  // 60
  read_two_bytes, // 70
  read_two_bytes, // 80
  read_two_bytes, // 90
  read_two_bytes, // a0
  read_two_bytes, // b0
  read_one_byte,  // c0
  read_two_bytes, // d0
  read_two_bytes, // e0
  read_two_bytes, // e0
};


static PyObject *
read_f0_byte (char **track, char *end, unsigned long time,
	      unsigned char x, unsigned char z)
	      
{
  debug_print ("%x:%s", z, "event\n");
  if (z == 0xff)
    {
      unsigned char zz = *(*track)++;
      debug_print ("%x:%s", zz, "f0-event\n");
      if (zz == 0x01 && **track <= 0x07)
	return read_string (track, end, time, x, zz);
      else if (zz == 0x2f && **track == 0x00)
	{
	  (*track)++;
	  debug_print ("%s", "done\n");
	  return 0;
	}
      else
	return read_unimplemented_event (track, end, time, x, zz);
      exit (0);
    }
  else
    return read_string (track, end, time, x, z);
}

Read_midi_event read_midi_event [16] =
{
  read_one_byte,  //  0
  read_one_byte,  // 10
  read_one_byte,  // 20
  read_one_byte,  // 30
  read_one_byte,  // 40
  read_one_byte,  // 50
  read_one_byte,  // 60
  read_two_bytes, // 70
  read_two_bytes, // 80
  read_two_bytes, // 90
  read_two_bytes, // a0
  read_two_bytes, // b0
  read_one_byte,  // c0
  read_two_bytes, // d0
  read_two_bytes, // e0
  read_f0_byte,   // f0
};
	     
static PyObject *
read_event (char **track, char *end, unsigned long time,
	    unsigned char *running_status)
{
  int rsb_skip = ((**track & 0x80)) ? 1 :0;

  unsigned char x = (rsb_skip) ? (*track)[0]: *running_status;
  // unsigned char y = x & 0xf0;
  unsigned char z = (*track)[1 + rsb_skip];
  
  debug_print ("%x:%s", z, "event\n");
  *track += 2 + rsb_skip;

  return (*read_midi_event[z >> 4]) (track, end, time, x, z);
}

static PyObject *
midi_parse_track (char **track, char *track_end)
{
  unsigned int time = 0;
  unsigned char running_status;
  unsigned long track_len, track_size;
  PyObject *pytrack = 0;

  debug_print ("%s", "\n");
  
  track_size = track_end - *track;
  /* need this for direct midi.parse_track (s) on midi file */
  if (strcmp (*track, "MTrk"))
    *track = memmem (*track, track_size - 1, "MTrk", 4);
  debug_print ("%s", "\n");
  assert (!strcmp (*track, "MTrk"));
  *track += 4;

  track_len = get_number (*track, *track + 4, 4);
  *track += 4;

  debug_print ("track_len: %u\n", track_len);
  debug_print ("track_size: %u\n", track_size);
  debug_print ("track begin: %p\n", track);
  debug_print ("track end: %p\n", track + track_len);
  
  //assert (track_len <= track_size);

  pytrack = PyList_New (0);

  track_end = *track + track_len;

  while (*track < track_end)
    {
      long dt = get_variable_length_number(track, track_end);
      time += dt;
      PyList_Append (pytrack, read_event (track, track_end, time,
					  &running_status));
    }

  *track = track_end;
  return pytrack;
}


static PyObject *
pymidi_parse_track (PyObject *self, PyObject *args)
{
  char *track, *track_end;
  unsigned long track_size, track_len;

  PyObject * sobj = PyTuple_GetItem (args, 0);

  debug_print ("%s", "\n");
  if (!PyArg_ParseTuple (args, "s#", &track, &track_size))
    return 0;

  if (track_size < 0)
    return midi_error ("negative track size");

  track_end = track + track_size;
  
  return midi_parse_track (&track, track_end);
}

static PyObject *
midi_parse (char **midi, char *midi_end)
{
  PyObject *pymidi = 0;
  unsigned long header_len;
  unsigned format, tracks;
  int division;
  int i;
  
  debug_print ("%s", "\n");

  /* Header */
  header_len = get_number (*midi, *midi + 4, 4);
  *midi += 4;
  
  if (header_len < 6)
    return midi_error ("header too short");
    
  format = get_number (*midi, *midi + 2, 2);
  *midi += 2;
  
  tracks = get_number (*midi, *midi + 2, 2);
  *midi += 2;

  if (tracks > 32)
    return midi_error ("too many tracks");
  
  division = get_number (*midi, *midi + 2, 2) * 4;
  *midi += 2;

  if (division < 0)
    /* return midi_error ("can't handle non-metrical time"); */
    ;
  *midi += header_len - 6;

  pymidi = PyList_New (0);
  PyList_Append (pymidi, Py_BuildValue ("(iii)", format, tracks, division));

  /* Tracks */
  for (i = 0; i < tracks; i++)
    PyList_Append (pymidi, midi_parse_track (midi, midi_end));

  return pymidi;
}

static PyObject *
pymidi_parse (PyObject *self, PyObject *args)
{
  char *midi, *midi_end;
  unsigned long midi_size, midi_len;
  
  PyObject *sobj = PyTuple_GetItem (args, 0);

  debug_print ("%s", "\n");
  if (!PyArg_ParseTuple (args, "s#", &midi, &midi_size))
    return 0;

  if (strcmp (midi, "MThd"))
      return midi_error ("MThd expected");
  
  midi += 4;

  midi_end = midi + midi_size;

  return midi_parse (&midi, midi_end);
}


static PyMethodDef MidiMethods[] = 
{
  {"parse",  pymidi_parse, 1},
  {"parse_track",  pymidi_parse_track, 1},
  {0, 0}        /* Sentinel */
};

initmidi ()
{
  PyObject *m, *d;
  m = Py_InitModule ("midi", MidiMethods);
  d = PyModule_GetDict (m);
  
  Midi_error = PyString_FromString ("midi.error");
  PyDict_SetItemString (d, "error", Midi_error);
  Midi_warning = PyString_FromString ("midi.warning");
  PyDict_SetItemString (d, "warning", Midi_warning);
}

File: python/midi.py
# This file is part of LilyPond, the GNU music typesetter.
#
# Copyright (C) 2001--2023 Han-Wen Nienhuys <hanwen@xs4all.nl>
#           Jan Nieuwenhuizen <janneke@gnu.org>
#
#
# LilyPond is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# LilyPond is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with LilyPond.  If not, see <http://www.gnu.org/licenses/>.

# import midi
# s = open ("s.midi").read ()
# midi.parse_track (s)
# midi.parse (s)
#
#
# returns a MIDI file as the tuple
#
#  ((format, division), TRACKLIST)          # division (>0) = TPQN*4
#                                           # or (<0) TBD
#
# each track is an EVENTLIST, where EVENT is
#
#   (time, (type, ARG1, [ARG2]))            # time = cumulative delta time
# MIDI event:
#   type = MIDI status+channel >= x80
# META-event  = xFF:
#   type = meta-event type <= x7F
#   ARG1 = length
#   ARG2 = data

import array
import struct


class error (Exception):
    pass

# class warning (Exception): pass


# TODO: should use enum?

# Channel voice messages
NOTE_OFF = 0x80
NOTE_ON = 0x90
POLYPHONIC_KEY_PRESSURE = 0xa0
CONTROLLER_CHANGE = 0xb0
PROGRAM_CHANGE = 0xc0
CHANNEL_KEY_PRESSURE = 0xd0
PITCH_BEND = 0xe0

# Channel mode messages
ALL_SOUND_OFF = 0x78
RESET_ALL_CONTROLLERS = 0x79
LOCAL_CONTROL = 0x7a
ALL_NOTES_OFF = 0x7b
OMNI_MODE_OFF = 0x7c
OMNI_MODE_ON = 0x7d
MONO_MODE_ON = 0x7e
POLY_MODE_ON = 0x7f

# Meta events
SEQUENCE_NUMBER = 0x0
TEXT_EVENT = 0x1
COPYRIGHT_NOTICE = 0x2
SEQUENCE_TRACK_NAME = 0x3
INSTRUMENT_NAME = 0x4
LYRIC = 0x5
MARKER = 0x6
CUE_POINT = 0x7
PROGRAM_NAME = 0x8
DEVICE_NAME = 0x9
MIDI_CHANNEL_PREFIX = 0x20
MIDI_PORT = 0x21
END_OF_TRACK = 0x2f
SET_TEMPO = 0x51
SMTPE_OFFSET = 0x54
TIME_SIGNATURE = 0x58
KEY_SIGNATURE = 0x59
XMF_PATCH_TYPE_PREFIX = 0x60
SEQUENCER_SPECIFIC_META_EVENT = 0x7f
META_EVENT = 0xff


def _get_variable_length_number(nextbyte, getbyte):
    sum = 0
    while nextbyte >= 0x80:
        sum = (sum + (nextbyte & 0x7F)) << 7
        nextbyte = getbyte()
    return sum + nextbyte


def _first_command_is_repeat(status, nextbyte, getbyte):
    raise error('the first midi command in the track is a repeat')


def _read_two_bytes(status, nextbyte, getbyte):
    return status, nextbyte


def _read_three_bytes(status, nextbyte, getbyte):
    return status, nextbyte, getbyte()


def _read_string(nextbyte, getbyte):
    length = _get_variable_length_number(nextbyte, getbyte)
    return ''.join(chr(getbyte()) for i in range(length))


def _read_f0_byte(status, nextbyte, getbyte):
    if status == 0xff:
        return status, nextbyte, _read_string(getbyte(), getbyte)
    return status, _read_string(nextbyte, getbyte)


_read_midi_event = (
    _first_command_is_repeat,  # 0
    None,  # 10
    None,  # 20
    None,  # 30
    None,  # 40
    None,  # 50
    None,  # 60 data entry???
    None,  # 70 all notes off???
    _read_three_bytes,  # 80 note off
    _read_three_bytes,  # 90 note on
    _read_three_bytes,  # a0 poly aftertouch
    _read_three_bytes,  # b0 control
    _read_two_bytes,  # c0 prog change
    _read_two_bytes,  # d0 ch aftertouch
    _read_three_bytes,  # e0 pitchwheel range
    _read_f0_byte,   # f0
)


def _parse_track_body(data, clocks_max):
    # This seems to be the fastest way of getting bytes in order as integers.
    dataiter = iter(array.array('B', data))
    getbyte = dataiter.__next__

    time = 0
    status = 0
    try:
        for nextbyte in dataiter:
            time += _get_variable_length_number(nextbyte, getbyte)
            if clocks_max and time > clocks_max:
                break
            nextbyte = getbyte()
            if nextbyte >= 0x80:
                status = nextbyte
                nextbyte = getbyte()
            yield time, _read_midi_event[status >> 4](status, nextbyte, getbyte)
    except StopIteration:
        # If the track ended just before the start of an event, the for loop
        # will exit normally. If it ends anywhere else, we end up here.
        print(len(list(dataiter)))
        raise error('a track ended in the middle of a MIDI command')


def _parse_hunk(data, pos, type, magic):
    if data[pos:pos+4] != magic:
        raise error('expected %r, got %r' % (magic, data[pos:pos+4]))
    try:
        length, = struct.unpack('>I', data[pos+4:pos+8])
    except struct.error:
        raise error(
            'the %s header is truncated (may be an incomplete download)' % type)
    endpos = pos + 8 + length
    data = data[pos+8:endpos]
    if len(data) != length:
        raise error(
            'the %s is truncated (may be an incomplete download)' % type)
    return data, endpos


def _parse_tracks(midi, pos, num_tracks, clocks_max):
    if num_tracks > 256:
        raise error('too many tracks: %d' % num_tracks)
    for i in range(num_tracks):
        trackdata, pos = _parse_hunk(midi, pos, 'track', b'MTrk')
        yield list(_parse_track_body(trackdata, clocks_max))
    # if pos < len(midi):
    #     warn


def parse_track(track, clocks_max=None):
    track_body, end = _parse_hunk(track, 0, 'track', b'MTrk')
    # if end < len(track):
    #     warn
    return list(_parse_track_body(track_body, clocks_max))


def parse(midi, clocks_max=None):
    header, first_track_pos = _parse_hunk(midi, 0, 'file', b'MThd')
    try:
        format, num_tracks, division = struct.unpack('>3H', header[:6])
    except struct.error:
        raise error('the file header is too short')
#  if division < 0:
#    raise error ('cannot handle non-metrical time')
    tracks = list(_parse_tracks(midi, first_track_pos, num_tracks, clocks_max))
    return (format, division*4), tracks
